% Generated by roxygen2 (4.1.1): do not edit by hand
% Please edit documentation in R/read.cds.R
\name{read.cds}
\alias{read.cds}
\title{Read the CDS of a given organism}
\usage{
read.cds(file, format, delete_corrupt = TRUE, ...)
}
\arguments{
\item{file}{a character string specifying the path to the file storing the CDS.}

\item{format}{a character string specifying the file format used to store the CDS, e.g. "fasta", "fatsq".}

\item{delete_corrupt}{a logical value indicating whether corrupt base triplets should be removed from the input \code{file}.}

\item{...}{additional arguments that are used by the \code{\link[Biostrings]{readDNAStringSet}} function.}
}
\value{
A data.table storing the gene id in the first column and the corresponding
sequence as string in the second column.
}
\description{
This function reads an organism specific CDS stored in a defined file format.
}
\details{
The \code{read.cds} function takes a string specifying the path to the cds file
of interest as first argument.

It is possible to read in different proteome file standards such as \emph{fasta} or \emph{fastq}.

CDS stored in fasta files can be downloaded from http://www.ensembl.org/info/data/ftp/index.html.
}
\examples{
\dontrun{
# reading a cds file stored in fasta format
Ath.cds <- read.cds(system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
                    format = "fasta")
}
}
\author{
Hajk-Georg Drost
}

